File: backend/app/models/task.py
"""
任务模型
"""
from datetime import datetime
from sqlalchemy import (
    Column,
    Integer,
    String,
    Text,
    DateTime,
    ForeignKey,
    Index,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship

from app.db.database import Base


class UnlockTask(Base):
    """解锁任务表"""

    __tablename__ = "unlock_tasks"

    id = Column(Integer, primary_key=True, index=True)
    account_id = Column(Integer, ForeignKey("apple_accounts.id"), nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)

    # 任务信息
    task_type = Column(String(50), nullable=False)  # unlock, disable_2fa, change_password, remove_devices
    task_id = Column(String(100), unique=True, nullable=True, index=True)  # Celery 任务 ID

    # 任务状态
    status = Column(String(50), default="pending", index=True)  # pending, processing, success, failed, cancelled
    priority = Column(Integer, default=0)

    # 执行信息
    node_id = Column(Integer, nullable=True)
    proxy_id = Column(Integer, nullable=True)

    # 结果
    result = Column(JSONB, nullable=True)
    error_message = Column(Text, nullable=True)
    retry_count = Column(Integer, default=0)
    max_retries = Column(Integer, default=3)

    # 时间信息
    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, index=True)

    # 关系
    account = relationship("AppleAccount", back_populates="tasks")

    # 索引
    __table_args__ = (
        Index("idx_tasks_status_created", "status", "created_at"),
        Index("idx_tasks_account_status", "account_id", "status"),
    )

    def __repr__(self):
        return f"<UnlockTask(id={self.id}, type='{self.task_type}', status='{self.status}')>"

File: frontend/src/api/task.ts
import { get, post } from '@/utils/request'

/**
 * 任务相关 API
 */

/**
 * 任务类型
 */
export interface UnlockTask {
  id: number
  user_id: number
  account_id: number
  task_type: 'unlock' | 'disable_2fa' | 'change_password'
  status: 'pending' | 'running' | 'success' | 'failed' | 'cancelled'
  priority: number
  node_id: number | null
  proxy_id: number | null
  started_at: string | null
  completed_at: string | null
  error_message: string | null
  retry_count: number
  created_at: string
  updated_at: string
}

/**
 * 任务列表参数
 */
export interface TaskListParams {
  page?: number
  size?: number
  status?: string
  task_type?: string
}

/**
 * 任务列表响应
 */
export interface TaskListResponse {
  items: UnlockTask[]
  total: number
  page: number
  size: number
}

/**
 * 创建任务请求
 */
export interface CreateTaskRequest {
  account_id: number
  task_type: 'unlock' | 'disable_2fa' | 'change_password'
  priority?: number
}

/**
 * 任务统计
 */
export interface TaskStats {
  total: number
  pending: number
  running: number
  success: number
  failed: number
  success_rate: number
  avg_duration: number
}

/**
 * 获取任务列表
 */
export function getTaskListApi(params?: TaskListParams) {
  return get<TaskListResponse>('/tasks', params)
}

/**
 * 获取任务详情
 */
export function getTaskDetailApi(id: number) {
  return get<UnlockTask>(`/tasks/${id}`)
}

/**
 * 创建解锁任务
 */
export function createUnlockTaskApi(data: CreateTaskRequest) {
  return post<UnlockTask>('/tasks/unlock', data)
}

/**
 * 取消任务
 */
export function cancelTaskApi(id: number) {
  return post(`/tasks/${id}/cancel`)
}

/**
 * 获取任务统计
 */
export function getTaskStatsApi() {
  return get<TaskStats>('/tasks/stats')
}
